File: processor/src/main/java/io/deloop/tools/reference/replay/Validator.java
package io.deloop.tools.reference.replay;

import com.google.auto.common.SuperficialValidation;
import io.deloop.tools.references.replay.ReplayAlways;

import javax.annotation.processing.Messager;
import javax.lang.model.element.Element;
import javax.lang.model.element.ElementKind;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Types;
import javax.tools.Diagnostic.Kind;
import java.util.ArrayList;
import java.util.List;

import static com.google.auto.common.MoreElements.asExecutable;
import static com.google.auto.common.MoreElements.isAnnotationPresent;
import static javax.lang.model.element.ElementKind.INTERFACE;
import static javax.lang.model.type.TypeKind.VOID;

class Validator {

    private final Element element;
    private final Types types;
    private final Messager messager;

    Validator(Element element, Types types, Messager messager) {
        this.element = element;
        this.types = types;
        this.messager = messager;
    }

    boolean validate() {
        return validateCompilable() &&
               validateInterface() &&
               validateOnlyVoidMethods() &&
               validateMaxOneReplayAlwaysMethod();
    }

    private boolean validateCompilable() {
        // compilation error in source code
        if (!SuperficialValidation.validateElement(element)) {
            warningCompilationErrors(element, messager);
            return false;
        }
        return true;
    }

    private boolean validateInterface() {
        if (!element.getKind().equals(INTERFACE)) {
            errorNotAnInterface();
            return false;
        }
        return true;
    }

    private boolean validateOnlyVoidMethods() {
        for (Element el : enclosedElements()) {
            if (isMethod(el) && !methodReturnsVoid(el)) {
                errorNoVoidMethod(el);
                return false;
            }
        }
        return true;
    }

    private boolean validateMaxOneReplayAlwaysMethod() {
        int count = 0;

        for (Element e : enclosedElements()) {
            if (isMethod(e) && isAnnotationPresent(e, ReplayAlways.class)) {
                count++;
            }
            if (count > 1) {
                errorTooManyReplayAlwaysMethods();
                return false;
            }
        }
        return true;
    }

    private void errorNotAnInterface() {
        String annotationName = ReplayAlways.class.getSimpleName();
        error(element, "Is not an interface. " + annotationName + " can only be applied to interfaces");
    }

    private void errorNoVoidMethod(Element method) {
        error(element, "Has non-void method " + method.getSimpleName());
    }

    private void errorTooManyReplayAlwaysMethods() {
        error(element, "Interface has too many methods annotated with" +
                ReplayAlways.class.getSimpleName() + ". Max is 1.");
    }

    private void error(Element e, String message) {
        messager.printMessage(Kind.ERROR, message, e);
    }

    private boolean isMethod(Element e) {
        return e.getKind().equals(ElementKind.METHOD);
    }

    private List<? extends Element> enclosedElements() {
        List<Element> result = new ArrayList<>();
        result.addAll(element.getEnclosedElements());

        // add methods of extended interfaces too
        for(TypeMirror tm : types.directSupertypes(element.asType())) {
            Element el = types.asElement(tm);
            if (el.getKind().equals(ElementKind.INTERFACE)) {
                result.addAll(el.getEnclosedElements());
            }
        }

        return result;
    }

    private boolean methodReturnsVoid(Element e) {
        return asExecutable(e).getReturnType().getKind().equals(VOID);
    }

    private static void warningCompilationErrors(Element element, Messager messager) {
        messager.printMessage(Kind.WARNING, "Had compilation errors", element);
    }
}

File: compiler/src/main/java/io/deloop/tools/reference/replay/Validator.kt
package io.deloop.tools.reference.replay

import com.google.auto.common.MoreElements.asExecutable
import com.google.auto.common.MoreElements.isAnnotationPresent
import com.google.auto.common.SuperficialValidation
import io.deloop.tools.references.replay.ReplayAlways
import io.deloop.tools.references.replay.ReplayReference
import java.lang.Exception
import java.util.*
import javax.annotation.processing.Messager
import javax.lang.model.element.Element
import javax.lang.model.element.ElementKind
import javax.lang.model.element.ElementKind.INTERFACE
import javax.lang.model.element.ElementKind.METHOD
import javax.lang.model.type.TypeKind.VOID
import javax.lang.model.util.Types
import javax.tools.Diagnostic.Kind

internal class Validator(private val element: Element, private val types: Types, private val messager: Messager) {

    fun validate() = try {
        checkIfCompiles()
        checkIfInterface()
        checkIfItHasOnlyVoidMethods()
        allowMaxOneReplayAlwaysMethod()

        true
    } catch (ignore: InvalidUsageException) {
        false
    }

    private fun checkIfCompiles() {
        if (!SuperficialValidation.validateElement(element)) {
            warningCompilationErrors(element, messager)
            fail()
        }
    }

    private fun checkIfInterface() {
        if (element.kind != INTERFACE) {
            val annotation = ReplayReference::class.java.simpleName
            error(element, "$annotation can only be applied to interfaces")
            fail()
        }
    }

    private fun checkIfItHasOnlyVoidMethods() {
        enclosedElements()
                .filter { isMethod(it) }
                .filter { !methodReturnsVoid(it) }
                .forEach {
                    error(element, "Has non-void method ${it.simpleName}")
                    fail()
                }
    }

    private fun allowMaxOneReplayAlwaysMethod() {
        enclosedElements()
                .filter { isMethod(it) }
                .filter { isAnnotationPresent(it, ReplayAlways::class.java) }
                .count()
                .apply {
                    if (this > 1) {
                        errorTooManyReplayAlwaysMethods()
                        fail()
                    }
                }
    }

    private fun errorTooManyReplayAlwaysMethods() {
        error(element, "Interface has too many methods annotated with" +
                ReplayAlways::class.java.simpleName + ". Max is 1.")
    }

    private fun error(e: Element, message: String) {
        messager.printMessage(Kind.ERROR, message, e)
    }

    private fun isMethod(e: Element) = e.kind == METHOD

    private fun enclosedElements(): List<Element> {
        val result = ArrayList<Element>()
        result.addAll(element.enclosedElements)

        // add methods of extended interfaces too
        for (tm in types.directSupertypes(element.asType())) {
            val el = types.asElement(tm)
            if (el.kind == ElementKind.INTERFACE) {
                result.addAll(el.enclosedElements)
            }
        }

        return result
    }

    private fun methodReturnsVoid(e: Element) = asExecutable(e).returnType.kind == VOID

    private fun warningCompilationErrors(element: Element, messager: Messager) {
        messager.printMessage(Kind.WARNING, "Had compilation errors", element)
    }

    private fun fail(): Nothing = throw InvalidUsageException()

    private class InvalidUsageException : Exception()
}
